### DISCUSSION TITLE
Store WebSocket user data in upgrade handler

### DISCUSSION THREAD

**Discussion by arthuro555** [uNetworking/uWebSockets.js#860] [Q&A] [answered]
Hello 👋 

I was wondering if it could be possible to store some user data in the upgrade handler.
I'd like to pass in a "roomID" parameters through the URL, verify if the user is allowed to join this room, and either block the establishment of a connection at this point, or, if the connection is allowed, store the roomID as a field of the UserData of the connection. 

I _guess_ I could do something like

```js
let roomID;
app.ws("/:roomID/", {
  upgrade: (res, req, context) => {
    roomID = req.getParameter(0);
    res.upgrade(
      { url: req.getUrl() },
      req.getHeader('sec-websocket-key'),
      req.getHeader('sec-websocket-protocol'),
      req.getHeader('sec-websocket-extensions'),
      context
    );
  },
  open: (ws) => {
    ws.getUserData().roomID = roomID;
  },
}); 
```

According to the `Upgrade` example, `open` should be called immediately when upgrading so this should be safe to do... But that still feels very wrong and unsafe 😓 Is there any way to set the userData directly in the upgrade handler?

**Comment by uNetworkingAB** [comment, marked as answer]
https://unetworking.github.io/uWebSockets.js/generated/interfaces/HttpResponse.html#upgrade

**Reply by arthuro555** [reply]
Ah, sorry for this 😅 I assumed it was just a weird way to pass the URL in and that the upgrade handler required it, when looking at the example code... Thanks for your time 🙏

**Reply by uNetworkingAB** [reply]
yeah I understood that - better if the example used "myData" instead of "url" to really highlight it being something by user feel free to PR example change

**Reply by uNetworkingAB** [reply]
https://github.com/uNetworking/uWebSockets.js/commit/4774d9242479dbdf0d25c93c4c2d6d5e71a04dc7
